examples/dagster_slurm_example/defs/shell/shell_integration.py

```diff
@@ -1,60 +1,301 @@
-import shutil
+import os
+import re
+import shlex
+import textwrap
+import time
+import uuid
+from contextlib import contextmanager
+from pathlib import Path
 
-import dagster as dg
+from dagster import (
+    AssetExecutionContext,
+    MaterializeResult,
+    PipesFileContextInjector,
+    PipesFileMessageReader,
+    asset,
+    open_pipes_session,
+)
+
+from .helpers import (
+    _detect_partition,
+    _dexec,
+    _find_controller,
+    _put_file,
+    _submit_python,
+    _tail_logs,
+    _wait_done,
+    upload_file,
+)
 
-from dagster_slurm_example.defs.shared import example_defs_prefix
+CONTAINER = "slurmctld"
+CONTAINER_DATA_DIR = "/data"
+HOST_SHARED = os.path.abspath("./shared")
+PIPES_BASE = os.path.join(HOST_SHARED, "pipes")
 
+LOCAL_PAYLOAD = os.environ.get(
+    "LOCAL_EXTERNAL_FILE",
+    str(
+        (
+            Path(__file__).resolve().parents[3]
+            / "dagster_slurm_example"
+            / "defs"
+            / "shell"
+            / "external_file.py"
+        ).resolve()
+    ),
+)
 
-def get_python_executable() -> str:
-    """Get the Python executable path from PATH.
+PYTHON_IN_NODE = os.environ.get("PYTHON_IN_NODE", "python3")
 
-    Returns:
-        str: Path to the Python executable
+PYTHON_IN_NODE = "/usr/bin/python3"
+DEFAULT_LOCAL_SCRIPT = os.path.join(os.path.dirname(__file__), "test.py")
+TERMINAL = {"COMPLETED", "FAILED", "CANCELLED", "TIMEOUT", "PREEMPTED", "NODE_FAIL"}
 
-    Raises:
-        RuntimeError: If Python executable is not found in PATH
-    """
-    python_path = shutil.which("python")
-    if python_path is None:
-        raise RuntimeError("Python executable not found in PATH")
-    return python_path
+
+def job_state(jid):
+    o, _, _ = _dexec(CONTAINER, f"squeue -h -j {jid} -o '%T' || true")
+    s = (o or "").strip()
+    if s:
+        return s
+    o, _, _ = _dexec(CONTAINER, f"sacct -X -n -j {jid} -o State || true")
+    s = (o or "").strip()
+    return s.split()[0] if s else ""
 
 
-@dg.multi_asset(
-    specs=[dg.AssetSpec(key=[example_defs_prefix, "orders"]), dg.AssetSpec("users")]
+@asset(
+    name="slurm_submit_pipes"
 )
-def subprocess_asset(
-    context: dg.AssetExecutionContext,
-    pipes_subprocess_client: dg.PipesSubprocessClient,
-):
-    python_path = get_python_executable()
-    cmd = [python_path, dg.file_relative_path(__file__, "shell_external.py")]
-
-    return pipes_subprocess_client.run(
-        command=cmd,
+def slurm_submit_remote_pipes(context: AssetExecutionContext):# noqa: C901
+    os.makedirs(PIPES_BASE, exist_ok=True)
+    _dexec(
+        CONTAINER,
+        "mkdir -p /data/{logs,results,pipes} && chmod 1777 /data/{logs,results,pipes}",
```

Comment:
As discussed today - we could write the logs to

- shared storage on the supercomputer
- S3-like (external)


For us s3 like would be more simple.

However, I think for others eventually using the library one more external dependency makes it harder to use it.

Can we standardize on using a in-cluster volume?

Reply:
If we intend to use the local library it would be an sdist and not a bdist as it is unlikely users would push that to pypi first. Is there a good way to handle that?

Reply:
https://github.com/Quantco/pixi-pack/issues/209

Reply:
shared library builds now and is automatically in `pixi run pack`